---
title: "Seurat_scRNAseq"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(Seurat)
library(ggplot2)
library(RColorBrewer)
library(cowplot)
library(scater)
library(sctransform)
library(tidyverse)

## working directory
dir <- ''

## data directory
dir.data <- '/hdd_scratch1/sam67253/cell_ranger-pipeline/data/'

```


```{r read_data, warning=FALSE, message=FALSE}

#Read in 10X raw data from mtx and tsv files for five samples

#Sample IDs - BRI-405, BRI-417, BRI-611, BRI-536, BRI-627. 


#for BRI-405
sample.ID <- 'BRI-405'
seurat_data <- Read10X(paste0(dir.data, sample.ID, '/filtered_feature_bc_matrix/'))
seurat_obj_BRI_405 <- CreateSeuratObject(counts = seurat_data$`Gene Expression`, project = sample.ID)
seurat_obj_BRI_405 <- RenameCells(object = seurat_obj_BRI_405, add.cell.id = sample.ID)


## For BRI-417
sample.ID <- 'BRI-417'
seurat_data <- Read10X(paste0(dir.data, sample.ID, '/filtered_feature_bc_matrix/'))
seurat_obj_BRI_417 <- CreateSeuratObject(counts = seurat_data$`Gene Expression`, project = sample.ID)
seurat_obj_BRI_417 <- RenameCells(object = seurat_obj_BRI_417, add.cell.id = sample.ID)


## For BRI-536
sample.ID <- 'BRI-536'
seurat_data <- Read10X(paste0(dir.data, sample.ID, '/filtered_feature_bc_matrix/'))
seurat_obj_BRI_536 <- CreateSeuratObject(counts = seurat_data$`Gene Expression`, project = sample.ID)
seurat_obj_BRI_536 <- RenameCells(object = seurat_obj_BRI_536, add.cell.id = sample.ID)


## For BRI-611
sample.ID <- 'BRI-611'
seurat_data <- Read10X(paste0(dir.data, sample.ID, '/filtered_feature_bc_matrix/'))
seurat_obj_BRI_611 <- CreateSeuratObject(counts = seurat_data$`Gene Expression`, project = sample.ID)
seurat_obj_BRI_611 <- RenameCells(object = seurat_obj_BRI_611, add.cell.id = sample.ID)


## For BRI-627
sample.ID <- 'BRI-627'
seurat_data <- Read10X(paste0(dir.data, sample.ID, '/filtered_feature_bc_matrix/'))
seurat_obj_BRI_627 <- CreateSeuratObject(counts = seurat_data$`Gene Expression`, project = sample.ID)
seurat_obj_BRI_627 <- RenameCells(object = seurat_obj_BRI_627, add.cell.id = sample.ID)

```

```{r}

# Creat a merged Seurat object
seurat_obj <- merge(seurat_obj_BRI_405,
                    c(seurat_obj_BRI_417, seurat_obj_BRI_536, seurat_obj_BRI_611, seurat_obj_BRI_627))
seurat_obj

str(seurat_obj)

# access raw counts from the Seurat object for the first twenty rows and columns
seurat_obj$RNA@counts[1:20, 1:30]

# access the meta data
head(seurat_obj@meta.data)

# create Seurat object
seurat_obj <- CreateSeuratObject(counts = seurat_data$`Gene Expression`)


```

```{r}

#Quality Control

#1. Remove poor quality barcodes. 

#Metrics for QC
#Unique genes detected, cell-doublets, no.of genes, count depth and mitochondrial expression

``` {r histogram_of_detected_genes, warning = FALSE, fig.width = 10, fig.height = 6}

hist <- ggplot(seurat_obj@meta.data, aes(x = nFeature_RNA)) +
  facet_wrap(~orig.ident) +
  geom_histogram(binwidth = 50) +
  xlim(0, max(seurat_obj@meta.data$nFeature_RNA) + 1000) +
  labs(x ='# detected genes') +
  theme_bw() +
  theme(plot.title = element_text(size = 14, face='bold'))

hist

```

```{r percentage_mitochondrial_genes}

# Calculate the Mitochondrial content for each single cell and add to meta.data

percent_mito <- PercentageFeatureSet(seurat_obj, pattern = '^MT-')
seurat_obj <- AddMetaData(object = seurat_obj,
                          metadata = percent_mito,
                          col.name = 'percent_mito')
```


```{r QC_violin_plots, warning = FALSE, message = FALSE, fig.width = 10, fig.height = 6}

#no. of detected genes per cell('nfeature_RNA'), no.of counts per #cell('nCount_RNA'), percent of mitochindiral genes per cell #('percent_mito')

violin_plt1 <- VlnPlot(seurat_obj, features = c('nFeature_RNA'), 
                       ncol = 1, pt.size = 0) +
  scale_y_continuous(trans = 'log10') +
  theme(legend.position = 'none')

violin_plt2 <- VlnPlot(seurat_obj, features = c('nCount_RNA'), 
                       ncol = 1, pt.size = 0) +
  scale_y_continuous(trans = 'log10') +
  theme(legend.position = 'none')

violin_plt3 <- VlnPlot(seurat_obj, features = c('percent_mito'), 
                       ncol = 1, pt.size = 0) +
  theme(legend.position = 'none')

plot_grid(violin_plt1, violin_plt2, violin_plt3, ncol = 2)

#the median no. of genes per cells is highest in 'BRI-405' and 'BRI-627'

```

#check correlation between the three QC metrics

```{r QC_scatter_plots, warning = FALSE, fig.width = 10, fig.height = 8}

feat_scatter1 <- FeatureScatter(seurat_obj, feature1 = 'nCount_RNA', feature2 = 'nFeature_RNA', 
                                group.by = 'orig.ident') +
  geom_hline(yintercept = 1000, linetype='dashed', colour = '#777777') +
  geom_vline(xintercept = 3000, linetype='dashed', colour = '#777777') +
  scale_x_continuous(trans = 'log10') +
  scale_y_continuous(trans = 'log10') +
  labs(x ='nCount_RNA', y = 'nFeature_RNA') +
  theme(legend.position = 'none')
feat_scatter2 <- FeatureScatter(seurat_obj, feature1 = 'nCount_RNA', feature2 = 'percent_mito', 
                                group.by = 'orig.ident') +
  geom_hline(yintercept = 20, linetype='dashed', colour = '#777777') +
  geom_vline(xintercept = 3000, linetype='dashed', colour = '#777777') +
  scale_x_continuous(trans = 'log10') +
  labs(x ='nCount_RNA', y = 'percent_mito') +
  theme(legend.position = 'none')
feat_scatter3 <- FeatureScatter(seurat_obj, feature1 = 'nFeature_RNA', feature2 = 'percent_mito', 
                                group.by = 'orig.ident') +
  geom_hline(yintercept = 20, linetype='dashed', colour = '#777777') +
  geom_vline(xintercept = 1000, linetype='dashed', colour = '#777777') +
  scale_x_continuous(trans = 'log10') +
  labs(x ='nFeature_RNA', y = 'percent_mito') +
  theme(legend.position = 'none')

```


#extract the legend


```{r}

legend <- get_legend(feat_scatter1 + theme(legend.position = 'right'))
plot_grid(feat_scatter2, feat_scatter3, feat_scatter1, legend, ncol = 2)

```

```{r}

#define outlier cells by median deviations

outliers.lib <- isOutlier(seurat_obj$nCount_RNA, type = 'lower', 
                          log = TRUE, nmads = 2, 
                          batch = seurat_obj$orig.ident)
outliers.ngenes <- isOutlier(seurat_obj$nFeature_RNA, type = 'lower', 
                             log = TRUE, nmads = 2, 
                             batch = seurat_obj$orig.ident)
outliers.mt <- isOutlier(seurat_obj$percent_mito, 
                         type = 'higher', nmads = 2, 
                         batch = seurat_obj$orig.ident)
thresholds <- data.frame(nCount_RNA = attr(outliers.lib, 'thresholds')[1,],
                         nFeature_RNA = attr(outliers.ngenes, 'thresholds')[1,],
                         percent_mito = attr(outliers.mt, 'thresholds')[2,])
round(thresholds, 2)

```

```{r bad_qual_cells}

#identify low quality samples 
badQual <- outliers.lib | outliers.ngenes | outliers.mt
table(remove=badQual)

```

```{r removal_bud_qual}

#remove bad quality samples from downstream processing.

seurat_obj <- seurat_obj[, !badQual]

#removes genes that are not expresses in any cell

seurat_obj <- seurat_obj[rowMeans(seurat_obj$RNA@counts) > 0, ]

seurat_obj

```

##Normalization


```{r sparsity_calc, warning = FALSE, fig.width = 5, fig.height = 3.5}

#normalize the raw counts. 


fraction_non_zero <- as.data.frame(colSums(seurat_obj$RNA@counts != 0) / dim(seurat_obj$RNA@counts)[1])
colnames(fraction_non_zero) <- c('fraction_non_zero')
p1 <- ggplot(fraction_non_zero) +
  geom_histogram(aes(x = fraction_non_zero), binwidth = 0.0025) +
  labs(title = 'sparsity of GEX data', x = 'fraction non-zero features', y = 'number of cells') +
  theme_classic()
p1

```


```{r sctransformation, warning = FALSE}
# run sctransform()

# Note: this step takes a bit of calculation time
seurat_obj <- SCTransform(seurat_obj, variable.features.n = 3000, verbose = FALSE)

```


##Highly variable genes to reduce the dimensionality of data.

```{r var_features, fig.width = 8, fig.height = 4}
seurat_obj$SCT@var.features[1:50]
ggplot(seurat_obj$SCT@meta.features) +
  geom_point(aes(x = sct.gmean, y = sct.residual_variance, color = sct.variable)) + 
  scale_color_manual(values=c('#999999', 'red')) +
  scale_x_continuous(trans='log10') +
  scale_y_continuous(trans='log10') +
  theme_bw()
```

##Dimenstionality reduction

```{r run_pca, fig.width = 10, fig.height = 5}

#PCA

seurat_obj <- RunPCA(seurat_obj, assay = 'SCT', verbose = FALSE, npcs = 50)
seurat_obj
str(seurat_obj@reductions$pca)
VizDimLoadings(seurat_obj, dims = 1:3, reduction = 'pca', ncol = 3)

```

```{r viz_pca, fig.width = 8, fig.height = 6}
DimPlot(seurat_obj, reduction = 'pca')
```


##Elbow plot

```{r elbow, fig.width = 10, fig.height = 4}

#to select the appropriate no.of dimesnions

ElbowPlot(seurat_obj, ndims = 25)
```


```{r tsne_umap, warning = FALSE, message = FALSE}
seurat_obj <- RunTSNE(seurat_obj, reduction = 'pca', verbose = FALSE, dims =  1:20)
seurat_obj <- RunUMAP(seurat_obj, n.components = 2L, reduction = 'pca', dims = 1:20, verbose = FALSE)
seurat_obj
str(seurat_obj@reductions)

```

##UMAP

```{r dim_vis_tsen_umap, fig.width = 10, fig.height = 5}
p1 <- DimPlot(seurat_obj, reduction = 'tsne')
p2 <- DimPlot(seurat_obj, reduction = 'umap')
plot_grid(p1, p2, ncol = 2)

```






